Add dbt clone operator

pankajastro opened the pull request:
## Description

This PR introduces the DbtCloneOperator. For more details, refer to the dbt documentation: https://docs.getdbt.com/reference/commands/clone.

## Testing

**Airflow DAG**
```python
from datetime import datetime

from airflow import DAG
from cosmos import DbtSeedLocalOperator, DbtRunLocalOperator, DbtCloneLocalOperator, ProfileConfig

DBT_PROJ_DIR="/usr/local/airflow/dbt/jaffle_shop"

profile_config1=ProfileConfig(
profile_name="bigquery_dev",
target_name="dev",
profiles_yml_filepath="/usr/local/airflow/dbt/jaffle_shop/profiles.yml"
)

profile_config2=ProfileConfig(
profile_name="bigquery_clone",
target_name="dev",
profiles_yml_filepath="/usr/local/airflow/dbt/jaffle_shop/profiles.yml"
)


with DAG("test-id-1", start_date=datetime(2024, 1, 1), catchup=False) as dag:
    seed_operator = DbtSeedLocalOperator(
        profile_config=profile_config1,
        project_dir=DBT_PROJ_DIR,
        task_id="seed",
        dbt_cmd_flags=["--select", "raw_customers"],
        install_deps=True,
        append_env=True,
    )
    run_operator = DbtRunLocalOperator(
        profile_config=profile_config1,
        project_dir=DBT_PROJ_DIR,
        task_id="run",
        dbt_cmd_flags=["--models", "stg_customers"],
        install_deps=True,
        append_env=True,
    )

    clone_operator = DbtCloneLocalOperator(
        profile_config=profile_config2,
        project_dir=DBT_PROJ_DIR,
        task_id="clone",
        dbt_cmd_flags=["--models", "stg_customers", "--state", "/usr/local/airflow/dbt/jaffle_shop/target"],
        install_deps=True,
        append_env=True,
    )

    seed_operator >> run_operator >> clone_operator
``` 

**DBT Profile**
```
bigquery_dev:
  target: dev
  outputs:
    dev:
      type: bigquery
      method: service-account
      project: astronomer-dag-authoring
      dataset: bq_dev
      threads: 4 # Must be a value of 1 or greater
      keyfile: /usr/local/airflow/include/key.json
      location: US


bigquery_clone:
  target: dev
  outputs:
    dev:
      type: bigquery
      method: service-account
      project: astronomer-dag-authoring
      dataset: bq_clone
      threads: 4 # Must be a value of 1 or greater
      keyfile: /usr/local/airflow/include/key.json
      location: US
``` 

**Airflow DAG Run**

<img width="1660" alt="Screenshot 2024-11-15 at 6 06 50 PM" src="https://github.com/user-attachments/assets/4a3af37e-3f6c-4859-814f-aeff3a252ac6">


**BQ data WH**

<img width="1454" alt="Screenshot 2024-11-15 at 6 04 29 PM" src="https://github.com/user-attachments/assets/69b45f57-3ff5-43d9-a8f1-bb04e7ad5735">


## Related Issue(s)

closes: https://github.com/astronomer/astronomer-cosmos/issues/1268
closes: https://github.com/astronomer/astronomer-cosmos/issues/878

## Breaking Change?
No

## Limitation
- The `dbt clone` command was introduced in dbt-core 1.6.0, so this feature is only available to users with dbt-core version 1.6 or higher https://github.com/dbt-labs/dbt-core/blob/1.6.latest/CHANGELOG.md
- Users should ensure their database is supported for cloning operations.

## Checklist

- [ ] I have made corresponding changes to the documentation (if required)
- [ ] I have added tests that prove my fix is effective or that my feature works

tatiana commented on cosmos/operators/aws_eks.py:
```diff
@@ -160,3 +161,12 @@ class DbtRunOperationAwsEksOperator(DbtAwsEksBaseOperator, DbtRunOperationKubern
         super().__init__(*args, **kwargs)
+
+
+class DbtCloneAwsEksOperator(DbtAwsEksBaseOperator, DbtCloneKubernetesOperator):
+    """
+    Executes a dbt core clone command.
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
```
What value is the constructor giving us? Would it make sense to do something like this:

```suggestion
class DbtCloneAwsEksOperator(DbtAwsEksBaseOperator, DbtCloneKubernetesOperator):
    """
    Executes a dbt core clone command.
    """
    pass
```

pankajastro replied:
@tatiana We introduced init in PR [#1162](https://github.com/astronomer/astronomer-cosmos/pull/1162) because the operator import issue was observed for Airflow version 2.10 + Cosmos, and [#1161](https://github.com/astronomer/astronomer-cosmos/issues/1161) was reported in Cosmos. Do you think we should keep this in this PR, and when we work on [#1161](https://github.com/astronomer/astronomer-cosmos/issues/1161), we can clean it up then? That way, we'll be more comfortable.

pankajastro replied:
There was an alternative solution proposed in PR https://github.com/astronomer/astronomer-cosmos/pull/1163, but it started breaking other tests. This may require some investigations.

tatiana replied:
Let's not block the current ticket due to this. I'm happy for us to review this approach in #1161 , as you suggested, @pankajastro !

tatiana commented on cosmos/operators/azure_container_instance.py:
```diff
@@ -167,3 +168,12 @@ class DbtRunOperationAzureContainerInstanceOperator(DbtRunOperationMixin, DbtAzu
+    Executes a dbt core clone command.
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any):
```
The same feedback given on https://github.com/astronomer/astronomer-cosmos/pull/1326/files#r1843904192 applies here and in the other `DbtClone` operators

tatiana reviewed (commented):
@pankajastro it's great to have you back!

Would it make sense to add the example DAG you used to test as one of Cosmos' integration DAGs?

Please, could you add documentation about this feature? I believe we may be lacking docs on how people can use `DbtBuild` operators - we could cover both in the same place.

Last but not least, it may be worth to have a test related to full refresh. Some Cosmos operators support full refresh, others don't. It seems clone supports this, so it would be great to make sure the interfaces are consistent with running models and seeds.

pankajastro commented:
> @pankajastro it's great to have you back!
> 
> Would it make sense to add the example DAG you used to test as one of Cosmos' integration DAGs?
> 
> Please, could you add documentation about this feature? I believe we may be lacking docs on how people can use `DbtBuild` operators - we could cover both in the same place.
> 
> Last but not least, it may be worth to have a test related to full refresh. Some Cosmos operators support full refresh, others don't. It seems clone supports this, so it would be great to make sure the interfaces are consistent with running models and seeds.

Thank you for the review!

I have added integration tests

For the scope of this PR, I have added docs for the clone operator.

I have updated existing unit tests for the full refresh, but please let me know if we need integration tests for this.
